src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -342,81 +348,95 @@ private static object GetTypeAsDiagnosticArgument(TypeSymbol typeOpt)
             return typeOpt ?? (object)"<null>";
         }
 
-        /// <summary>
-        /// Report nullable mismatch warnings and optionally update tracked value on assignment.
-        /// </summary>
-        private void TrackNullableStateForAssignment(BoundNode node, int targetSlot, TypeSymbolWithAnnotations targetType, BoundExpression value, TypeSymbolWithAnnotations valueType, int valueSlot)
+        private void ReportAssignmentWarnings(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, bool useLegacyWarnings)
         {
             Debug.Assert(!IsConditionalState);
             if (this.State.Reachable)
             {
-                if ((object)targetType == null)
+                if (targetType is null || valueType is null)
                 {
                     return;
                 }
 
-                if (targetType.IsReferenceType)
+                if (targetType.IsReferenceType && targetType.IsNullable == false && valueType.IsNullable == true)
                 {
-                    bool isByRefTarget = IsByRefTarget(targetSlot);
-
-                    if (targetType.IsNullable == false)
+                    if (useLegacyWarnings)
                     {
-                        if (valueType?.IsNullable == true && (value == null || !CheckNullAsNonNullableReference(value)))
+                        ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_ConvertingNullableToNonNullable, value.Syntax);
+                    }
+                    else
+                    {
+                        if (!CheckNullAsNonNullableReference(value))
                         {
-                            ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullReferenceAssignment, (value ?? node).Syntax);
+                            ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullReferenceAssignment, value.Syntax);
```

Comment:
Should we return if something was reported, or still continue to `ReportNullabilityMismatchInAssignmentIfNecessary` below?

Reply:
We could but this check is for top-level nullability and the check below is for nested nullability. The two seem independent but I'm on the fence whether we should suppress the second warning.

---
In reply to: [175196964](https://github.com/dotnet/roslyn/pull/25516#discussion_r175196964) [](ancestors = 175196964)

Reply:
OK to leave as-is. We'll see from experience.